[WIP] Quick fix for inconsistent SDK versions

brodybits opened the pull request:
Resolves #508 - Inconsistent handling of min/target SDK values

This is a quick fix to use the minimum & target SDK version values from AndroidManifest.xml.

I have tested it in a newly generated Cordova project from the command line.

This proposal also resolves #631 - Target SDK version not consistent with defaults in root build.gradle bug

~~TODO items:~~

- [x] ~~update unit tests to succeed with the new code change~~
- [x] ~~update unit tests to cover extracting minimum & target SDK version values from AndroidManifest.xml~~

Unfortunately this does NOT resolve the issue reported in #629 (errors encountered on Android Studio 3.2 & 3.3).

I really hope this kind of a solution will be part of the upcoming major release for Cordova 9 (apache/cordova#10).

fredgalvao commented on bin/templates/cordova/lib/builders/ProjectBuilder.js:
```diff
@@ -89,17 +89,29 @@ class ProjectBuilder {
-        return packageName.substring(lastDotIndex + 1);
+        const m2 = /<uses-sdk[\s\S]*?android:minSdkVersion\s*=\s*"(.*?)"/i.exec(manifestData);
+        if (!m2) {
+            throw new CordovaError('Could not android:minSdkVersion in ' + manifestPath);
```
Was this supposed to be `Could not <find> android:min...`? The same applies to the `target` one bellow.

brodybits replied:
Now fixed.

fredgalvao commented on bin/templates/cordova/lib/builders/ProjectBuilder.js:
```diff
@@ -89,17 +89,29 @@ class ProjectBuilder {
+        }
+
+        const m3 = /<uses-sdk[\s\S]*?android:targetSdkVersion\s*=\s*"(.*?)"/i.exec(manifestData);
+        if (!m2) {
```
Was this `if` supposed to look for `!m3` instead?

brodybits replied:
> Was this `if` supposed to look for `!m3` instead?

Now fixed.

fredgalvao commented on bin/templates/cordova/lib/builders/ProjectBuilder.js:
```diff
@@ -144,7 +162,21 @@ class ProjectBuilder {
+            '$1=' + androidManifestInfo.targetSdkVersion + ' //');
+        rootGradle = rootGradle.replace(
+            /(defaultCompileSdkVersion)\s*=\s*\d+\s*\/\//,
+            '$1=' + androidManifestInfo.targetSdkVersion + ' //');
```
Just to confirm, we're using the `targetSdkVersion` as the `defaultCompileSdkVersion`, right?

brodybits replied:
Comment added

brodybits commented on bin/templates/cordova/lib/builders/ProjectBuilder.js:
```diff
@@ -89,17 +89,29 @@ class ProjectBuilder {
+
+        const m3 = /<uses-sdk[\s\S]*?android:targetSdkVersion\s*=\s*"(.*?)"/i.exec(manifestData);
+        if (!m2) {
+            throw new CordovaError('Could not android:targetSdkVersion in ' + manifestPath);
```
> Missing "find" here as well (adding this comment so it does not get forgotten)

Now fixed

brodybits commented:
Updated as follows:

- Updated ProjectBuilder.spec.js according to changes, with additional tests to cover reading of `android:minSdkVersion` and `android:targetSdkVersion` values from `AndroidManifest.xml`
- use `parseInt` on `android:minSdkVersion` and `android:targetSdkVersion` values
- fix use of wrong variable in an if statement, as discussed in review comments
- add missing "find" in a couple error message, as discussed in review comments

erisu commented:
@brodybits Can you take a look into the AppVeyor build failure?

brodybits commented:
Now green after an empty commit

brodybits commented:
Thanks @erisu for the review. I would consider this to be a stopgap or makeshift solution since it does not solve #508 and #631 the best way possible.

I would much rather find a way to resolve #508 and #631 *without* injecting the minimum & default SDK values into `AndroidManifest.xml` as a side effect. I think injecting the values into `AndroidManifest.xml` implies that the settings in `AndroidManifest.xml` are still relevant, which is not true if these settings are otherwise configured in Gradle.

An additional problem was already reported in #629: if we inject these SDK values in `AndroidManifest.xml`, recent versions of Android Studio would complain that these values should not be defined in a manifest file and offer to update manifest and Gradle files. This is behavior we would rather that our users do not see.

There was discussion in #508 about using `GradlePropertiesParser.js` to set the values in Gradle based on the settings in `config.xml`. I wonder if there would be an easy way to do this?

In case we do continue with this PR, I would definitely favor adding the `@todo` comment and changing "quick fix" to "stopgap" in all places including title, commit message, and comments.

The motivation for this stopgap solution was to solve the problem without blocking the new release for Cordova 9.

I just marked this proposal as [WIP] again, pending discussion whether to continue or find a better way to solve the problem.

brodybits commented:
Closing in favor of alternatives discussed in <https://github.com/apache/cordova-android/issues/508#issuecomment-461672598> and <https://github.com/apache/cordova-android/issues/508#issuecomment-461679701>
